backend/controller/UserController.js

```diff
@@ -0,0 +1,77 @@
+const { User } = require('../schema/User');
+
+/**
+ * The user's ID is autogenerated as the field _id.
+ *
+ * @param {*} req.body.name full name of a user
+ * @param {*} req.body.token auth token
+ */
+exports.createUser = (req, res) => {
+  const user = new User(req.body);
+  user.save((err, user) => {
+    if (err) {
+      console.error(err);
+      res.send(err);
+    } else res.send(user);
+  });
+};
+
+/**
+ * @param {*} req.params.id: the user id
+ */
+exports.findUserById = (req, res) => {
+  const id = req.params.id;
+  User.findById(id, (err, user) => {
+    if (err) {
+      console.error(err);
+      res.send(err);
+    } else res.send(user);
+  });
+};
+
+/**
+ * This IS AND SHOULD be a PATCH request. It is not a PUT request because
+ * it is not designed to replace an entire entity.
+ *
+ * Even though $addToSet and $pull apply mutations to a collection, this
+ * function is idempotent. Applying the same request multiple times will
+ * yield the same result as the initial application:
+ *
+ * $addToSet https://docs.mongodb.com/manual/reference/operator/update/addToSet/
+ * - For a given field, add value(s) if it does not already exist.
+ *
+ * $pull https://docs.mongodb.com/manual/reference/operator/update/pull/
+ * - For a given field, remove all instances of value(s).
+ *
+ * example form:
+ * {
+ *   'name: 'Some edited name',
+ *   'addToSet: {
+ *     'friendIds': 'friend-id-to-add',
+ *     'topSongIds': 'song-id-to-add'
+ *   },
+ *   'pull: {
+ *     'friendIds': 'friend-id-to-remove',
+ *     'topSongIds': 'song-id-to-remove'
+ *   },
+ * }
+ *
+ * @param {*} req.params.id the user id
+ * @param {*} req.body.{any} modified value of a userSchema field
+ * @param {*} req.body.addToSet use this to add friendIds and topSongIds
+ * @param {*} req.body.pull use this to remove friendIds and topSongIds
+ */
+exports.updateUserById = (req, res) => {
+  const id = req.params.id;
+  const { addToSet, pull, ...fieldsToUpdate } = req.body;
+  if (pull) fieldsToUpdate.$pull = JSON.parse(pull);
+  if (addToSet) fieldsToUpdate.$addToSet = JSON.parse(addToSet);
```

Comment:
what is the `$` syntax

Reply:
It's an update operator that denotes a set of actions. Without `$`, MongoDB will think it's a `key: value` pair to be updated.